linux/src/threadpool_linux.c

```diff
@@ -501,93 +490,125 @@ int threadpool_schedule_work(THANDLE(THREADPOOL) threadpool, THREADPOOL_WORK_FUN
     return result;
 }
 
-int threadpool_timer_start(THANDLE(THREADPOOL) threadpool, uint32_t start_delay_ms, uint32_t timer_period_ms, THREADPOOL_WORK_FUNCTION work_function, void* work_function_ctx, TIMER_INSTANCE_HANDLE* timer_handle)
+static void threadpool_timer_dispose(THREADPOOL_TIMER * timer)
 {
-    int result;
+    /* Codes_SRS_THREADPOOL_LINUX_07_071: [ threadpool_timer_dispose shall call timer_delete to destroy the ongoing timers. ]*/
+    if (timer_delete(timer->time_id) != 0)
+    {
+        LogErrorNo("Failure calling timer_delete.");
+    }
+    else
+    {
+        // Do Nothing
+    }
+    // Even though timer_delete does cancel any events, there is a small window where an event was triggered just before and a thread is being created.
+    // So the callback can still execute after the timer_delete call. A small wait here keeps the timer instance alive long enough for
+    // the callback to use it if we hit this window.
+
+    /* Codes_SRS_THREADPOOL_LINUX_07_096: [ threadpool_timer_dispose shall call ThreadAPI_Sleep to allow timer resources to clean up. ]*/
+    ThreadAPI_Sleep(30);
```

Comment:
>30 [](http://example.com/codeflow?start=20&length=2)

I noticed this got changed from 10 to 30.
Isn't our code good enough now to not have this at all?
If this Sleep is still needed, we should file a task under Things that never get done to get rid of it.

My suspicion is that we do not need it any longer. I know David put this one in so would be great to check briefly with him and if not needed any longer I would love to just remove it!

Reply:
I worked with David about this issue together last week. I saw your comment about better to remove this pause in previous iteration, so David and I tried different approaches to see if this pause can be removed but all failed so add this back. I added a task under things never get done to later get rid of it [https://msazure.visualstudio.com/One/_backlogs/backlog/Azure%20Messaging%20Store/Backlog%20items/?workitem=31375231](https://msazure.visualstudio.com/One/_backlogs/backlog/Azure%20Messaging%20Store/Backlog%20items/?workitem=31375231)